### PR TITLE
[file-format] add friend visibility

### PR DESCRIPTION
**Pull request by mengxu-fb**
## Motivation

This PR aims to add finer-grained visibility control to the Move language on top of the existing `public` and `private` visibility modifier. To be specific, given a function `f`, three access modifiers can be specified at function definition:
- **public**: functions defined in any module / script can call function `f`
- **private**: only functions defined in the same module can call function `f`
- **[IN THIS PR] public(friend)**: functions defined in the same module + *friend modules* (i.e., a controlled list of entities) can call function `f`

### Added language syntax and semantics:

There will be two additions added to the language in the first stage of implementation:
1. A friend list, specifiable at each module, that lists the modules that are considered as a *friend* of the host module.
    - Currently, only modules *within the same address as the host module* can be added to the friend list.
2. A protected keyword, specifiable at each function, to indicate the function can be callable from "friend" modules.

In summary, any function defined in the modules in the **friend list** can call any **protected** function defined in the host module.

In the following example:

```
0x2 {
  module A {
    friend 0x2::B;                      --> A friend list declared for each module
    friend 0x2::C;

    public(friend) fun foo() {}         --> A new visibility modifier: public(friend)
    public(friend) fun bar() {}

    fun i_am_private() {}
    public fun i_am_public() {}
  }
}
```

- Any function defined in host module (`0x2::A`) or friend modules (`0x2::B`, `0x2::C`) can call `foo` and `bar`, but no other functions can.
- Only functions defined in `0x2::A` can call `i_am_private`.
- Any function / script can call `i_am_public`.
- Only modules owned by `0x2` can be placed in the friend list.

### Progress tracking

This PR will be confined to the file format + basic checks (e.g., bounds check and duplication check) associated with the new file format. Other TODO items will come in follow-up PRs.

- Bytecode format
  - [DONE] versioned deserializer
  - [DONE] file format changes
  - [DONE] serialization of the new format with visibility info
  - [DONE] deserialization of both old and new versions of the bytecode
  - [DONE] valid friend list generation strategy (for proptest)
  - [TODO] more tests, especially on differences of the two versions of bytecode format
- Bytecode verifier
  - [DONE] bounds check of the new `CompiledModule` + tests + invalid mutations
  - [DONE] duplication check on the friend list + tests
  - [DONE] dependency check on whether a callsite can call into a `public(friend)` function
  - [TODO] existence of friend modules
  - [TODO] validity of friend declaration (e.g., creating cyclic dependencies is not allowed)
  - [TODO] same-address rule in friend declaration
  - [TODO] module upgrade compatibility checking
- IR compiler
  - [TODO] syntax and parsing for friend list and the protected modifier
  - [TODO] validation and dependency passing
  - [TODO] test cases on the new visibility feature
- Source compiler
  - [DONE] interface file generation with `public(friend)` functions
  - [TODO] syntax and parsing for friend list and the protected modifier
  - [TODO] validation and dependency passing
  - [TODO] test cases on the new visibility feature
- Move prover
  - [TODO] changes to `move-model`
- Diem framework
  - [TODO] re-write some of the modules with the new *protected* keyword
- Workflow
  - [WIP] finalize on the module publishing / re-publishing flow
  - [WIP] fix remaining issues related single-module republishing
  - [TODO] single-signer multi-module publishing in one transaction 
  - [TODO] E2E test cases

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

`cargo xtest`

New test cases will be added and existing tests will be updated per each commit. We will try to keep each commit small and sensible while maintaining a passing status of each commit.

## Related PRs

This PR depends on #7323 authored by @tnowacki
Another related development effort is VM republishing flow, tracked by #7143 and #7234.

### CODE REVIEW

File: language/testing-infra/e2e-tests/goldens/language_e2e_testsuite::tests::data_store::borrow_after_move.exp
Comment (mengxu-fb) [inline language/testing-infra/e2e-tests/goldens/language_e2e_testsuite::tests::data_store::borrow_after_move.exp:1]: E2E tests golden files updated because the version number in the `CompiledModule` is bumped from 1 to 2.

File: language/vm/src/file_format.rs
Comment (tnowacki) [inline language/vm/src/file_format.rs:368]: We can probably remove this struct?

File: language/vm/src/file_format.rs
Comment (tnowacki) [inline language/vm/src/file_format.rs:1651]: I'm not super convinced that we should be using the term "friend" any more. I think at the bytecode level we should just be more explicit and descriptive. Pulling into a single term like "friend" seems useful only if we need to refer to this in a bunch of different places. Which I don't think is true?

So perhaps just 
```
pub protected_access: Vec<ModuleHandleIndex>
```

File: language/vm/src/file_format.rs
Comment (tnowacki) [reply language/vm/src/file_format.rs:368]: I'm guessing you added this because of the TODO of more policies, but there isn't really any point to having this struct now. And we will have to bump the fileformat version and do a refactor for any extension 

So best to not have the layer of indirection here if we don't need it

**Review by tnowacki — COMMENTED** [review]
In this "progress tracking" you have a big section for each major component. Do you plan on doing all of that on this PR? (please don't :))

File: language/vm/src/file_format.rs
Comment (mengxu-fb) [reply language/vm/src/file_format.rs:368]: Yes, the main reason for a struct here is for the ease of extensibility. But yes, the struct is not really needed now.

File: language/vm/src/file_format.rs
Comment (sblackshear) [inline language/vm/src/file_format.rs:346]: I think it is redundant/error-prone to have both this enum and the `u8` constants. I would kill the `enum` and stick to the constants, since that's the style used elsewhere in the file format. The reason it is done this way: using raw bytecodes makes it easier to see what the binary format is, whereas using Rust-specific features like enums can obscure that.

File: language/vm/src/file_format.rs
Comment (mengxu-fb) [reply language/vm/src/file_format.rs:1651]: I am fine with either naming, but I guess we may actually need refer to the "friend" concept across the stack? like file format, bytecode verifier, IR compiler, source compiler, prover, VM loader, etc?

Comment (sblackshear) [reply language/vm/src/file_format.rs:1651]: I strongly agree--I think `friend` is a bad term for us to use because it implies symmetry, but the language feature is not symmetric. Perhaps we should elevate this terminology question to a broader forum, but I think something like "client" or "patron" that connotes anti-symmetry would be good.

**Comment by mengxu-fb** [comment]
> In this "progress tracking" you have a big section for each major component. Do you plan on doing all of that on this PR?

Actually, I am not sure what is the best way forward:
1) having a separate commit for each functionality but group all commits under this same PR
2) having smaller PRs (like this one, it is self-contained, and all tests are passing), and merge them separately.

Thoughts?

**Comment by sblackshear** [comment]
I think (2) is much easier to review + is less likely to result in merge conflicts.

File: language/vm/src/file_format.rs
Comment (tnowacki) [reply language/vm/src/file_format.rs:1651]: Agree with Sam, and I would go one step further: I think we don't need to pin ourselves to a term like "friend" or "client" at all because the usecase is narrow. 

My vote for the syntax is 
```
protected  { 0x1::A, 0x2::B, 0x3::C }
```
No need for another keyword and it makes it hopefully more clear what is going on. If I saw "friend" here I would have no initial guess that it relates to `protected`

File: language/vm/src/deserializer.rs
Comment (tnowacki) [inline language/vm/src/deserializer.rs:1216]: Probably fine to just use 1 and 2 here. I don't see the constant doing much :)

File: language/vm/src/file_format.rs
Comment (mengxu-fb) [reply language/vm/src/file_format.rs:346]: Not sure I entirely get this. When speaking about the file format, there may be two things we are talking about:
1. The `CompiledModule`, which is known in the Rust world. In `CompiledModule`, the `Visibility` is an enum,
2. The serialized bytes, which is what is stored on-chain. In the raw bytes, the `Visibility` is a bit in the u8 flag.

A similar situation is the handling of `Kind {All, Resource, Copyable}`. It is an enum in the Rust world, and a `u8` flag in the serialized bytes.

File: language/vm/src/deserializer.rs
Comment (tnowacki) [inline language/vm/src/deserializer.rs:1238]: I would rewrite this as 
```
let is_protected = (flags & FunctionDefinition::PROTECTED) != 0;
let is_public = (flags & FunctionDefinition::PUBLIC) != 0;
match (is_protected, is_public) {
  (false, false) => Private,
  (true, false) => Protected,
  (false, true) => Public,
  (true, true0 => malformed
}
```
I find this easier to case through than nested ifs

File: language/vm/src/file_format.rs
Comment (tnowacki) [reply language/vm/src/file_format.rs:346]: I partially agree, and ran into a similar issue with constraints (working on an initial PR for that but was blocked by the versioning thing)

The benefits for the enum here are huge for writing code that checks visibility, granted that code is really rather small. 

I think you could maybe do 
```
#[repr(u8)]
pub enum Visibility {
  Private = 0x0,
  Public = 0x1,
  Protected = 0x2,
}
```
And then use these in the serializer and deserializer. 

So I guess I agree with sam that it is bad to have an enum and the u8 constants, but it is okay if the enum is the same as the u8 constants. It's when you have two things that things go wrong.

Comment (tnowacki) [reply language/vm/src/file_format.rs:346]: The flags case is a bit different, but this is how we represent bytecodes (with an repr'd enum and then another bag of opcode constants)

**Comment by tnowacki** [comment]
Definitely (2). 

Luckily you can stub a lot of things out with protected as any other code can just fail when it encounters it for now

File: language/vm/src/file_format.rs
Comment (tnowacki) [reply language/vm/src/file_format.rs:346]: Yeah I think deserializing flags into enums is probably fine, especially since they are ensured to be exclusive in the deserializer, putting it not into an enum would feel misleading. 

But I think where `Kind` didn't do the best was having the u8 representation totally different from the enum

File: language/vm/src/file_format.rs
Comment (tnowacki) [reply language/vm/src/file_format.rs:368]: Yeah I don't see it making it easier to making things extensible in the future. As in, the amount of work to actually extend anything isn't reduced by this struct

File: language/vm/src/file_format.rs
Comment (mengxu-fb) [reply language/vm/src/file_format.rs:346]: Got it! Thanks for the description and the suggestion! I think we can go with this:

```
#[repr(u8)]
pub enum Visibility {
  Private = 0x0,
  Public = 0x1,
  Protected = 0x2,
}
```

Comment (tnowacki) [reply language/vm/src/file_format.rs:346]: I wouldn't get too eager and switch to that just yet :) I'd like to see if that jives with @sblackshear's concerns

**Comment by mengxu-fb** [comment]
Cool, option (2) -- small and incremental PRs it is then :)

File: language/vm/src/file_format.rs
Comment (sblackshear) [reply language/vm/src/file_format.rs:346]: I think the enum + `repr` is a great approach. My main concern was indeed about the redundant representation.

File: language/testing-infra/e2e-tests/goldens/language_e2e_testsuite::tests::data_store::borrow_after_move.exp
Comment (mengxu-fb) [reply language/testing-infra/e2e-tests/goldens/language_e2e_testsuite::tests::data_store::borrow_after_move.exp]: With 13661c2, not only the version number is bumped to 2, the `flags` byte is changed as well.

File: language/vm/src/file_format.rs
Comment (mengxu-fb) [reply language/vm/src/file_format.rs:347]: Updated with 13661c2. Now the bytecode format is a bit more different:

Originally, the `flags` byte has two bits in use:
- 0x1: is_public
- 0x2: is_native

With the new format, the `flags` byte now essentially becomes two bytes:
- byte 1: visibility (with 0 - private, 1 - public, 2 - protected, and in the future, 3 - script, 4 - package, etc...).
- byte 2: now only bit 0x2 is used to indicate whether this is_native.

File: language/vm/src/deserializer.rs
Comment (mengxu-fb) [reply language/vm/src/deserializer.rs:1112]: To me having a constant imported from somewhere feels a bit safer... But yeah, purely personal preference.

File: language/vm/src/file_format.rs
Comment (mengxu-fb) [reply language/vm/src/file_format.rs:385]: After some refactoring trials I still want to keep the struct even if there is only one field here. It kinds of gives me a mentality that I am dealing with a "friend" instead of just a ModuleId, and we just happen to define that the unit of "friend" is a module.

But I also see the point that it may be good to ease the concept of "friend" out from the bytecode level, especially since we haven't fully decided on the name, so the fewer "friend" in var/struct names, the better.

Tough choice.

File: language/vm/src/deserializer.rs
Comment (tnowacki) [reply language/vm/src/deserializer.rs:1112]: I'm not sure how much safer it is if VERSION_1 is always just 1?

Maybe it conveys a bit more intent? I could see that maybe, but I think `version() < 2` has been fairly clear in my usecase

Comment (mengxu-fb) [reply language/vm/src/deserializer.rs:1112]: Yeah, I think VERSION_1 is probably going to be 1 all the time, and same for VERSION_2. Maybe the usecase I was thinking about is, say, I want to look at the places where there is a special handling of VERSION_n, I could do a `grep VERSION_n` to pinpoint the places.

Comment (tnowacki) [reply language/vm/src/deserializer.rs:1112]: Yeah thats not a bad one. Let's do the constants then

File: language/vm/src/file_format.rs
Comment (mengxu-fb) [reply language/vm/src/file_format.rs:1660]: After some internal discussion, we decided to go with the `friend` keyword and using `public(friend)` as the visibility modifier.

Comment (tnowacki) [reply language/vm/src/file_format.rs:1660]: Do we want `friend` for `friends`? :P 

Seriously, I don't know. 

I think `public(friend)` reads better. But `friends { 0x1::A, 0x2::B }` reads better

Comment (mengxu-fb) [reply language/vm/src/file_format.rs:1660]: I think `public(friend)` reads better and aligns with `public(script)`.

An alternative of friend declaration would be
```
friend 0x1::A;
friend 0x2::B;
```
similar to how we deal with `use`.

Comment (bob-wilson) [reply language/vm/src/file_format.rs:1660]: I'll vote for the plural. `public(friends)` reads better to me. If you compare it to Rust's `pub(crate)`, the code is only in one crate so it makes sense to use singular `crate` -- the visibility is public within the one crate. In our case, `public(friends)` means that the function is public to friend modules, and since there can be multiple friends, `public(friends)` is the right way to say that.

Comment (bob-wilson) [reply language/vm/src/file_format.rs:1660]: ...but after reading Meng's response, I would also feel OK about the singular `friend`. The comparison to `public(script)` is better than my comparison to Rust's `pub(crate)`

Comment (tnowacki) [reply language/vm/src/file_format.rs:1660]: AH! very clever! I like it :) 

We should maybe add coalescing to this and to use. So you can do `friend 0x2::{A, B};`

We should do this for use too. But its okay if it doesn't make it in the first PR

**Comment by mengxu-fb** [comment]
Rebased past #7534, merge conflicts with script visibility resolved.

**Comment by mengxu-fb** [comment]
I addressed most comments in this PR except for whether we want to keep the single-field struct `FriendDeclaration`.

```rust
pub struct FriendDeclaration {
    /// Handle index to the friend module
    pub module: ModuleHandleIndex,
}
```

After some refactoring trials I still want to keep the struct even if there is only one field here. It kinds of gives me a mentality that I am dealing with a "friend" instead of just a `ModuleId`, and we just happen to define that the unit of "friend" is a module.

File: language/bytecode-verifier/src/dependencies.rs
Comment (tnowacki) [inline language/bytecode-verifier/src/dependencies.rs:101]: if-let + else is not my friend :P 

Either switch to a match, or `self_module.map(|self_id| friend_module_ids.contains(self_id)).unwrap_or(false)`

File: language/compiler/ir-to-bytecode/src/compiler.rs
Comment (tnowacki) [inline language/compiler/ir-to-bytecode/src/compiler.rs:568]: Maybe tag yourself in it? Not needed, but I find that helpful sometimes. I do `// TODO(tmn)`

Regardless, a short description is nice incase this is missed and someone comes and reads the code later

File: language/move-model/src/model.rs
Comment (tnowacki) [inline language/move-model/src/model.rs:2405]: nit, I might have done this for the script case... but you can do `unimplemented!(..)` which gives a *slightly* better message when it is hit

File: language/vm/src/file_format.rs
Comment (tnowacki) [inline language/vm/src/file_format.rs:349]: We should remove this struct as the encapsulation is not needed

File: language/vm/src/file_format.rs
Comment (tnowacki) [inline language/vm/src/file_format.rs:1662]: Nit, I'd move this up with the other tables. It is weird for it to come after them (probably put it right after field handles)

File: language/move-vm/runtime/src/unit_tests/vm_arguments_tests.rs
Comment (tnowacki) [inline language/move-vm/runtime/src/unit_tests/vm_arguments_tests.rs:208]: be sure to move this when you move the field in the declaration order

File: language/vm/src/compatibility.rs
Comment (tnowacki) [inline language/vm/src/compatibility.rs:95]: We should totally rewrite this. I can put up another PR after this lands as you work on other things

File: language/vm/src/deserializer.rs
Comment (tnowacki) [inline language/vm/src/deserializer.rs:513]: if binary.version() < Version_2 ?

File: language/vm/src/deserializer.rs
Comment (tnowacki) [inline language/vm/src/deserializer.rs:1189]: I imagine Rust will CSE this, but maybe  move `u64::from(table.count)` to a local binding? Ignore this if none of the other `load_<table>` code does this

File: language/vm/src/file_format.rs
Comment (tnowacki) [inline language/vm/src/file_format.rs:1607]: be sure to move this when you move the field in the declaration order

File: language/vm/src/file_format.rs
Comment (tnowacki) [inline language/vm/src/file_format.rs:1741]: be sure to move this when you move the field in the declaration order

File: language/compiler/ir-to-bytecode/src/compiler.rs
Comment (mengxu-fb) [reply language/compiler/ir-to-bytecode/src/compiler.rs:568]: It's actually already done, but yes, I can elaborate the TODO and tag myself.

File: language/vm/src/compatibility.rs
Comment (mengxu-fb) [reply language/vm/src/compatibility.rs:95]: I have re-written this in https://github.com/diem/diem/pull/7531/commits/d518d19e907ea64e5ca36ae615824f1c03e84186.

File: language/vm/src/deserializer.rs
Comment (mengxu-fb) [reply language/vm/src/deserializer.rs:1189]: Other `load_*` seem to be doing this as well, like `load_identifiers` and `load_constant_pool`. So I guess this should be OK.

Comment (tnowacki) [reply language/vm/src/deserializer.rs:1189]: I imagine Rust is smart enough :) Just something to consider

File: language/vm/src/file_format.rs
Comment (mengxu-fb) [reply language/vm/src/file_format.rs:1662]: Since we are moving this, I think we can put it right after `module_handles` and before  `struct_handles`? The rationale is that the friend list is expected to be declared before any functions and structs are defined.

Comment (mengxu-fb) [reply language/vm/src/file_format.rs:1662]: Ignore the comment above. Yes, this list should be after the handles. After `field_handles` is the right place.

File: language/vm/src/file_format.rs
Comment (mengxu-fb) [reply language/vm/src/file_format.rs]: `FriendDeclaration` dropped.

**Review by tnowacki — APPROVED** [review]
Thanks for all the iterations! Looks good!

**Pull request closed without merging**
